### PR TITLE
[DNM] mon, cephfs: Add auth caps for CephFS fsids

### PR DESCRIPTION
**Pull request by fullerdj**
Add new mon and mds auth caps to restrict access based on fsid.

DNM for now because there is no test yet.

### CODE REVIEW

File: PendingReleaseNotes [hunk 1]
```diff
@@ -0,0 +1,8 @@
+14.1.1
+------
+
+* The monitor cluster may now withhold filesystem and/or MDS information from
+  clients based on a new auth cap:
+  [mon] allow rw; allow fsid=1; allow fsid=2
+* Communication with MDS daemons may now be restricted using a new auth cap:
+  [mds] allow fsid=1
```
Comment (batrick) [inline PendingReleaseNotes]: Like `mds allow rw fsid=1`?

File: PendingReleaseNotes [hunk 2]
```diff
@@ -0,0 +1,8 @@
+14.1.1
+------
+
+* The monitor cluster may now withhold filesystem and/or MDS information from
+  clients based on a new auth cap:
+  [mon] allow rw; allow fsid=1; allow fsid=2
```
Comment (batrick) [inline PendingReleaseNotes]: IIRC, the caps are additive. Wouldn't `allow rw` provide access to all MDSMaps?

File: doc/cephfs/client-auth.rst [hunk 3]
```diff
@@ -146,3 +146,45 @@ The optional ``{network/prefix}`` is a standard network name and
...
+``Path restriction`` above). To restrict communication to MDS daemons
+associated with a particular filesystem (or filesystems), use:
+
+::
+
+ client.0
+     key: AQAz7EVWygILFRAAdIcuJ12opU/JKyfFmxhuaw==
+     caps: [mds] allow rw, allow fsid=1
+     caps: [mon] allow r, allow fsid=1
+     caps: [osd] allow rw tag cephfs data=cephfs_a
```
Comment (batrick) [inline doc/cephfs/client-auth.rst]: I guess we also need to extend these OSDCaps to take an FSCID. Not urgent though.

File: src/mon/Monitor.cc [hunk 4]
```diff
@@ -2998,8 +3005,14 @@ void Monitor::get_cluster_status(stringstream &ss, Formatter *f)
 	mgrmon()->get_map().print_summary(nullptr, &ss);
 	ss << "\n";
       }
-      if (mdsmon()->get_fsmap().filesystem_count() > 0) {
-	ss << "    mds: " << spacing << mdsmon()->get_fsmap() << "\n";
+      const FSMap *fsmapp = &mdsmon()->get_fsmap();
```
Comment (batrick) [inline src/mon/Monitor.cc]: nit: `const auto& fsmap` would be better.

File: src/mon/Monitor.cc [hunk 5]
```diff
@@ -2998,8 +3005,14 @@ void Monitor::get_cluster_status(stringstream &ss, Formatter *f)
...
+      if (!fsids.empty()) {
+	FSMap map = *fsmapp;
```
Comment (batrick) [inline src/mon/Monitor.cc]: then just: `FSMap map = fsmap`.

File: doc/cephfs/client-auth.rst [hunk 6]
```diff
@@ -146,3 +146,23 @@ The optional ``{network/prefix}`` is a standard network name and
...
+available. In this case, the monitor cluster will only inform clients about
+filesystems specified by the administrator. Other filesystems will not be
+reported and commands affecting them will be ignored as if the filesystems
+did not exist.
+
+::
+
+ client.0
+     key: AQAz7EVWygILFRAAdIcuJ12opU/JKyfFmxhuaw==
+     caps: [mon] allow r, allow fsid=1, allow fsid=2
```
Comment (liewegas) [inline doc/cephfs/client-auth.rst]: ditto.  the first allow r applies to all fsids.  the fsid= has to be a property of each allow to restrict it.

File: PendingReleaseNotes (hunk 1 above)
Comment (fullerdj) [reply PendingReleaseNotes]: This implementation doesn't have a read-only variant. If you're restricted by fsid, the mds will drop all messages for unauthorized fsids. Do we want a read-only variant?

File: doc/cephfs/client-auth.rst (hunk 3 above)
Comment (fullerdj) [reply doc/cephfs/client-auth.rst]: Well, the OSD does not know the fscid. We could tag the pools with it, though.

**Comment by batrick** [comment]
ping @fullerdj

**Comment by fullerdj** [comment]
ack, @batrick . I should be able to get back on this next week.

File: PendingReleaseNotes (hunk 1 above)
Comment (batrick) [reply PendingReleaseNotes]: > This implementation doesn't have a read-only variant.

I think I was wondering if `mds allow r fsid=1` would be allowed. That is, read-only MDS access although I'm not sure who uses that in practice; would that imply a read-only mount?

File: doc/cephfs/client-auth.rst (hunk 3 above)
Comment (batrick) [reply doc/cephfs/client-auth.rst]: The pools tags are good enough now I think. I don't believe we want to move forward anymore with shared data pools. The feature is no longer attractive with the pg_autoscaler.

**Comment by fullerdj** [comment]
closed in favor of #32581

**Pull request closed without merging**